packages/shared/lib/sdk/sync.ts

```diff
@@ -307,6 +339,173 @@ export class NangoAction {
 
         return this.attributes as A;
     }
+
+    public async *paginate<T = any>(config: ProxyConfiguration): AsyncGenerator<T[], undefined, void> {
+        const providerConfigKey: string = this.providerConfigKey as string;
+        const template: Template = configService.getTemplate(providerConfigKey);
+        const templatePaginationConfig: Pagination | undefined = template.proxy?.paginate;
+
+        if (!templatePaginationConfig) {
+            throw Error(`Pagination is not supported for '${providerConfigKey}'. Please, add pagination config to 'providers.yaml' file`);
+        }
+
+        let paginationConfig: Pagination = templatePaginationConfig;
+        delete paginationConfig.limit;
+
+        if (config.paginate) {
+            const paginationConfigOverride: Record<string, any> = config.paginate as Record<string, any>;
+
+            if (paginationConfigOverride) {
+                paginationConfig = { ...paginationConfig, ...paginationConfigOverride };
+            }
+        }
+
+        if (!config.method) {
+            config.method = 'GET';
+        }
+
+        const configMethod: string = config.method.toLocaleLowerCase();
+        const passPaginationParamsInBody: boolean = ['post', 'put', 'patch'].includes(configMethod);
+
+        const updatedBodyOrParams: Record<string, any> = ((passPaginationParamsInBody ? config.data : config.params) as Record<string, any>) ?? {};
+        const limitParameterName: string = paginationConfig.limit_parameter_name;
+
+        if (paginationConfig['limit']) {
+            updatedBodyOrParams[limitParameterName] = paginationConfig['limit'];
+        }
+
+        // TODO: Consider creating 'Paginator' interface and moving the case block to specific implementations of 'Paginator'
+        switch (paginationConfig.type) {
+            case PaginationType.CURSOR: {
+                const cursorPagination: CursorPagination = paginationConfig as CursorPagination;
+
+                let nextCursor: string | undefined;
+                while (true) {
+                    if (nextCursor) {
+                        updatedBodyOrParams[cursorPagination.cursor_parameter_name] = nextCursor;
+                    }
+
+                    this.updateConfigBodyOrParams(passPaginationParamsInBody, config, updatedBodyOrParams);
+
+                    const response: AxiosResponse = await this.proxy(config);
+
+                    const responseData: T[] = cursorPagination.response_data_path
+                        ? this.getNestedField(response.data, cursorPagination.response_data_path)
+                        : response.data;
+
+                    if (!responseData.length) {
+                        return;
+                    }
+
+                    yield responseData;
+
+                    nextCursor = this.getNestedField(response.data, cursorPagination.next_cursor_parameter_path);
+
+                    if (!nextCursor || nextCursor.trim().length === 0) {
```

Comment:
Might be silly, but do we want to interrupt pagination if we get less objects than the limit? Same for the other pagination types?

Reply:
- For cursor-based pagination, we don't need to rely on limit since once we receive a response with smaller object count, we also get empty or no cursor in that same response.
- Same ☝️ is true for link. There won't be any if there are less records.
- We can introduce this as a small optimization for offset-based pagination. We won't be able to always do this because we rely on default API endpoint limit, which we don't know. So, the optimization will only kick in if users explicitly pass `limit` override from the script.
- This is dangerous approach for non-offset-based API in general because some of them specifically suggest that they could return a smaller number of objects in a batch. That's why they strongly suggest relying on their defined pagination mechanism instead of relying on the count of objects. [Here](https://api.slack.com/methods/conversations.list) is an example of this from Slack docs.
<img width="1139" alt="image" src="https://github.com/NangoHQ/nango/assets/17253295/19be10c8-93f2-4128-a3d1-1f2ebddfc885">

Reply:
Understood, this sounds reasonable to me!